arch/configure_reader.py

```diff
@@ -0,0 +1,628 @@
+#!/usr/bin/env python3
+
+import argparse
+import sys
+import os
+import re
+import inspect
+import platform
+from shutil import which
+
+archBlock  = re.compile( r"(?:#[ ]*)(ARCH(?:.*\n)*?)(?:#{5,})", re.I )
+kvPair     = re.compile( r"^(\w+)(?:[ \t]*=[ \t]*)(.*?)$",      re.I | re.M )
+# Make this gnarly and complicated since configure.defaults has no standard formatting
+#                            v start  v OS      V typical   v MACOS
+osAndArch     = re.compile( r"^ARCH[ ]+(\w+)[ ]+((?:\w+.*?),|(?:[(].*?[)]))",         re.I )
+# Just grab the first two words, thats what you get
+osAndArchAlt  = re.compile( r"^ARCH[ ]+(\w+)[ ]+(\w+)",         re.I )
+
+referenceVar  = re.compile( r"[$]([(])?(\w+)(?(1)[)])", re.I )
+compileObject = re.compile( r"(\W)-c(\W)" )
+
+class Stanza():
+  
+  def __init__( self, lines ) :
+    self.lines_   = lines
+    self.os_      = None
+    self.arch_    = None
+    self.osArchLine_ = None
+    self.archs_   = []
+    self.kvPairs_ = {}
+    self.crossPlatform_ = False
+    self.skipCrossPlatform_ = True
+    self.serialOpt_  = False
+    self.smparOpt_   = False
+    self.dmparOpt_   = False
+    self.dmsmOpt_    = False
+
+  def parse( self ) :
+    self.osArchLine_ = self.lines_.partition("\n")[0]
+    # First get os & archs
+    osarchMatch = osAndArch.match( self.osArchLine_ )
+
+    if osarchMatch is None :
+      osarchMatch = osAndArchAlt.match( self.osArchLine_ )
+      if osarchMatch is None :
+        print( "Could not find OS and architecture info in " + self.osArchLine_ )
+      
+    self.os_    = osarchMatch.group(1)
+    self.archs_ = osarchMatch.group(2).strip(",").split( " " )
```

Comment:
This RE breaks on macOS, because it reads the (MACOS) comment from the `configure.defaults` file. You could either make a special-case check here, or, update `configure.defaults` with something like this:

```diff
diff --git a/arch/configure.defaults b/arch/configure.defaults
index 36ca1b6e..34038a87 100644
--- a/arch/configure.defaults
+++ b/arch/configure.defaults
@@ -1000,7 +1000,7 @@ CC_TOOLS        =      $(SCC)
 NETCDFPAR_BUILD        =      CONFIGURE_NETCDFPAR_BUILD
 
 ###########################################################
-#ARCH    Darwin (MACOS) gfortran with gcc #serial smpar dmpar dm+sm
+#ARCH    Darwin x86_64, aarch64 (MACOS) gfortran with gcc #serial smpar dmpar dm+sm
 #
 DESCRIPTION     =       GNU ($SFC/$SCC)
 DMPARALLEL      =       # 1
@@ -1045,7 +1045,7 @@ CC_TOOLS        =      $(SCC)
 NETCDFPAR_BUILD        =      CONFIGURE_NETCDFPAR_BUILD
 
 ###########################################################
-#ARCH    Darwin (MACOS) gfortran with clang #serial smpar dmpar dm+sm
+#ARCH    Darwin x86_64, aarch64 (MACOS) gfortran with clang #serial smpar dmpar dm+sm
 #
 DESCRIPTION     =       GNU ($SFC/clang)
 DMPARALLEL      =       # 1
```

Reply:
I don't think I follow. The `osAndArch` regex properly catches the lines you mentioned in my tests when enabling (via changing the skip logic) cross-platform stanza selection. The last bit of the regex has a non-capture group for the `(...)` to deal with `(MACOS)`

It would be nice though to not have to deal with special cases for all the inconsistent formatting in comments/descriptions of things in `configure.defaults`

Reply:
On my MacBook Pro, the script failed to find any stanzas until I manually added the architectures to the `configure.defaults` file. Printing out the `Stanza.archs_` variable is how I found that it was capturing the `(MACOS)` token. I did not try using the cross-platform flag.  

I also found that using the gfortran/clang stanza failed as well because the `-fallow-argument-mismatch` flag was being passed to both gfortran and clang (where it's an invalid argument). Using gcc/gfortran worked fine after the edits to the configure file. I'll dig into that clang issue and post another comment when I determine the underlying issue.

Reply:
I see. I think I was focused on simply getting the regex to capture something, but that something is eventually compared to `platform.machine()` in python. I think a change would be best.

Reply:
I tried again but got the same result (no stanzas) when building on a M1 Mac -- as it turns out, macOS uses `arm64` as an architecture identifier vs `aarch64` that Linux uses. My patch worked on a Intel Mac but not Arm-based. My apologies; I should have double-checked that.

Reply:
Just double checking, adding `arm64` to the list of archs in the line does properly show the modified stanza?

Reply:
> Just double checking, adding `arm64` to the list of archs in the line does properly show the modified stanza?

Yes, but I would change `aarch64` to `arm64` for the Darwin stanzas vs. adding it. The macOS reports it only as `arm64` or `x86_64` from the `uname()` call (which is what Python is using under the hood).

Reply:
Okay, should be fixed now